File: server/src/main/webapp/WEB-INF/rails/app/controllers/api/web_hooks/guess_url_web_hook_controller.rb
#
# Copyright 2019 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Api
  module WebHooks
    class GuessUrlWebHookController < WebHookController

      protected
      def possible_urls
        %W(
          https://#{repo_host_name}/#{repo_full_name}
          https://#{repo_host_name}/#{repo_full_name}/
          https://#{repo_host_name}/#{repo_full_name}.git
          https://#{repo_host_name}/#{repo_full_name}.git/
          http://#{repo_host_name}/#{repo_full_name}
          http://#{repo_host_name}/#{repo_full_name}/
          http://#{repo_host_name}/#{repo_full_name}.git
          http://#{repo_host_name}/#{repo_full_name}.git/
          git://#{repo_host_name}/#{repo_full_name}
          git://#{repo_host_name}/#{repo_full_name}/
          git://#{repo_host_name}/#{repo_full_name}.git
          git://#{repo_host_name}/#{repo_full_name}.git/
          git@#{repo_host_name}:#{repo_full_name}
          git@#{repo_host_name}:#{repo_full_name}/
          git@#{repo_host_name}:#{repo_full_name}.git
          git@#{repo_host_name}:#{repo_full_name}.git/
          ssh://#{repo_host_name}/#{repo_full_name}
          ssh://#{repo_host_name}/#{repo_full_name}/
          ssh://#{repo_host_name}/#{repo_full_name}.git
          ssh://#{repo_host_name}/#{repo_full_name}.git/
          ssh://git@#{repo_host_name}/#{repo_full_name}
          ssh://git@#{repo_host_name}/#{repo_full_name}/
          ssh://git@#{repo_host_name}/#{repo_full_name}.git
          ssh://git@#{repo_host_name}/#{repo_full_name}.git/
        )
      end

      def repo_log_name
        "#{repo_host_name}/#{repo_full_name}"
      end

      def repo_host_name
        raise 'Subclass responsibility!'
      end

      def repo_full_name
        raise 'Subclass responsibility!'
      end
      
    end
  end
end

File: api/api-webhook-v1/src/main/java/com/thoughtworks/go/apiv1/webhook/GuessUrlWebHookController.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv1.webhook;

import spark.Request;

import java.util.List;

import static java.lang.String.format;

public interface GuessUrlWebHookController {

    String repoHostName(Request request) throws Exception;

    String repoFullName(Request request);

    default List<String> possibleUrls(Request request) throws Exception {
        String repoHostName = repoHostName(request);
        String repoFullName = repoFullName(request);
        return List.of(
            format("https://%s/%s", repoHostName, repoFullName),
            format("https://%s/%s/", repoHostName, repoFullName),
            format("https://%s/%s.git", repoHostName, repoFullName),
            format("https://%s/%s.git/", repoHostName, repoFullName),
            format("http://%s/%s", repoHostName, repoFullName),
            format("http://%s/%s/", repoHostName, repoFullName),
            format("http://%s/%s.git", repoHostName, repoFullName),
            format("http://%s/%s.git/", repoHostName, repoFullName),
            format("git://%s/%s", repoHostName, repoFullName),
            format("git://%s/%s/", repoHostName, repoFullName),
            format("git://%s/%s.git", repoHostName, repoFullName),
            format("git://%s/%s.git/", repoHostName, repoFullName),
            format("git@%s:%s", repoHostName, repoFullName),
            format("git@%s:%s/", repoHostName, repoFullName),
            format("git@%s:%s.git", repoHostName, repoFullName),
            format("git@%s:%s.git/", repoHostName, repoFullName),
            format("ssh://git@%s/%s", repoHostName, repoFullName),
            format("ssh://git@%s/%s/", repoHostName, repoFullName),
            format("ssh://git@%s/%s.git", repoHostName, repoFullName),
            format("ssh://git@%s/%s.git/", repoHostName, repoFullName)
        );
    }
}
